### ISSUE TITLE
Enhancement Request for Conditional Backend Builds and Build Dependencies in Amplify Gen 2

### ISSUE THREAD

**Issue by chelorope** [aws-amplify/amplify-hosting#3945]
### Before opening, please confirm:

- [X] I have checked to see if my question is addressed in the [FAQ](https://github.com/aws-amplify/amplify-hosting/blob/master/FAQ.md).
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-hosting/issues?q=is%3Aissue+).
- [X] I have removed any sensitive information from my code snippets and submission.

### Amplify Hosting feature

Backend builds

### Is your feature request related to a problem? Please describe:

Yes, my feature request is related to a problem I'm currently experiencing with Amplify Gen 2. I have configured my project as a monorepo, where an independent app is used to deploy backend resources, and two other apps are used to deploy different frontends. As decrived in the following article: https://docs.amplify.aws/react/deploy-and-host/fullstack-branching/monorepos/
The issue arises when I push any changes to the repository. Even if these changes are unrelated to the backend code, the backend app always triggers a build.

In Amplify Gen 1, there was an environment variable called AMPLIFY_DIFF_BACKEND that allowed the system to check for changes in the backend code and only trigger the backend build if changes were detected. This feature is not available in Amplify Gen 2, leading to unnecessary backend builds for every push to the repository, regardless of whether the changes are relevant to the backend.

This problem causes another issue when sharing backend resources with the frontend apps. To share these resources, I need to run npx ampx generate outputs --branch dev app-id AWS_APP_ID. This command works well unless the backend app is building simultaneously, in which case I receive the error: "Error: This branch deployment is in progress. Re-run this command once the deployment completes." Since the backend app deployment currently runs for every change pushed to the frontend code, this error occurs frequently, disrupting the development workflow.

### Describe how you'd like this feature to work

I would like Amplify Gen 2 to introduce a feature similar to the AMPLIFY_DIFF_BACKEND environment variable that was available in Amplify Gen 1. This feature should check for changes in the backend code and trigger the backend build only if changes are detected. This would prevent unnecessary builds and improve the efficiency of the build process.

Additionally, I would like a feature that allows one app to wait for another app to finish building before it starts its build process. This feature would work similarly to dependencies used in GitHub Actions, ensuring that dependent builds do not interfere with each other. For example, if the backend app is building, the frontend apps should wait for the backend build to complete before starting their build processes. This would prevent the error "Error: This branch deployment is in progress. Re-run this command once the deployment completes" and streamline the development workflow.

**Comment by Engineersticity** [comment]
Hey @chelorope (and anyone else running into this), 

I was dealing with this exact same pain point in Gen 2. Every frontend tweak was triggering a heavy CloudFormation backend build, and if I bypassed it, the frontend lost its outputs.

To solve this, I built an open-source CLI wrapper that recreates the old AMPLIFY_DIFF_BACKEND functionality but specifically tailors it for Gen 2 and monorepos.

Simply put `npx skip-amplify-backend` in front of your deploy command in amplify.yml:

`npx --yes skip-amplify-backend npx ampx pipeline-deploy --branch $AWS_BRANCH --app-id $AWS_APP_ID`

Why it fixes the issue:

It does a local git diff on the backend folder (you can set SKIP_AMPLIFY_FOLDER="apps/backend" for your monorepos).
If it detects zero changes, it securely exits the deploy command AND automatically runs `npx ampx generate outputs` so your frontend build doesn't crash from missing backend resources.

You can grab it here if it helps unblock your workflow: https://www.npmjs.com/package/skip-amplify-backend
